src/dune_engine/package.ml

```diff
@@ -295,22 +297,36 @@ module Source_kind = struct
       | Github -> "github.com"
       | Bitbucket -> "bitbucket.org"
       | Gitlab -> "gitlab.com"
+      | Sourcehut -> "sr.ht"
 
-    let homepage { kind; user; repo } =
+    let base_uri { kind; user; repo } =
       let host = host_of_kind kind in
-      sprintf "https://%s/%s/%s" host user repo
+      sprintf "%s/%s/%s" host
+        (match kind with
+        | Sourcehut -> "~" ^ user
+        | _ -> user)
+        repo
+
+    let add_https s = "https://" ^ s
+
+    let homepage t = add_https (base_uri t)
 
     let bug_reports t =
-      homepage t
-      ^
       match t.kind with
-      | Bitbucket | Github -> "/issues"
-      | Gitlab -> "/-/issues"
+      | Sourcehut -> add_https ("todo." ^ base_uri t)
+      | _ -> (
+        homepage t
+        ^
+        match t.kind with
+        | Sourcehut -> assert false
+        | Bitbucket | Github -> "/issues"
+        | Gitlab -> "/-/issues")
```

Comment:
The current implementation is fine, but I wonder if this would be even simpler to inline the intermediate abstractions and do a "plain" pattern matching. There will be some duplication, but the various hosts we support are already slightly different.

Reply:
There's a way to factor this stuff out better most certainly. I did not try very hard I admit.